crates/solver/src/liquidity/zeroex.rs

```diff
@@ -159,33 +155,37 @@ impl LiquidityCollecting for ZeroExLiquidity {
     }
 }
 
-fn generate_order_buckets(
-    zeroex_orders: Vec<OrderRecord>,
-    relevant_pairs: HashSet<TokenPair>,
-) -> OrderBuckets {
-    // divide orders in buckets
-    let mut buckets = OrderBuckets::default();
-    zeroex_orders
-        .into_iter()
-        .filter(
-            |record| match TokenPair::new(record.order.taker_token, record.order.maker_token) {
-                Some(pair) => relevant_pairs.contains(&pair),
-                None => false,
-            },
-        )
-        .for_each(|order| {
-            let bucket = buckets
-                .entry((order.order.taker_token, order.order.maker_token))
-                .or_default();
-            bucket.push(order);
-        });
-    buckets
+fn group_by_token_pair(
+    orders: impl Iterator<Item = OrderRecord>,
+) -> HashMap<(H160, H160), Vec<OrderRecord>> {
+    orders
+        .filter_map(|record| {
+            TokenPair::new(record.order().taker_token, record.order().maker_token).map(|_| {
+                (
+                    (record.order().taker_token, record.order().maker_token),
+                    record,
+                )
+            })
+        })
+        .into_group_map()
 }
 
 /// Get the `orders_per_type` best priced and biggest volume orders.
-fn get_useful_orders(order_buckets: OrderBuckets, orders_per_type: usize) -> Vec<OrderRecord> {
+fn get_useful_orders(
+    order_buckets: &OrderBuckets,
+    relevant_pairs: HashSet<TokenPair>,
+    orders_per_type: usize,
+) -> Vec<OrderRecord> {
     let mut filtered_zeroex_orders = vec![];
-    for mut orders in order_buckets.into_values() {
+    for orders in order_buckets
+        .iter()
+        .filter_map(|((token_a, token_b), record)| {
+            TokenPair::new(*token_a, *token_b)
+                .is_some_and(|pair| relevant_pairs.contains(&pair))
+                .then_some(record)
```

Comment:
The relevant pairs (<=13) will likely be fewer than all the buckets so iterating over the relevant pairs and indexing into the `order_buckets` should be faster and nicer to read.
This also shows that the key of the buckets should now probably be `TokenPair` instead of `(H160, H160)`.

Reply:
> This also shows that the key of the buckets should now probably be TokenPair instead of (H160, H160).

Well, the TokenPair is sorted inside. Previously the logic was the following:
```
        .filter(|record| {
            match TokenPair::new(record.order().taker_token, record.order().maker_token) {
                Some(pair) => relevant_pairs.contains(&pair),
                None => false,
            }
        })
        .for_each(|order| {
            let bucket = buckets
                .entry((order.order().taker_token, order.order().maker_token))
                .or_default();
            bucket.push(order);
        })
```

Initially, I used the TokenPair as a key but reverted the change since the tests started to fail.

Reply:
IMO, altering the logic might be worth a separate PR.

Reply:
Yeah, that's fair. 👌